runtime/runtime/src/bandwidth_scheduler/scheduler.rs

```diff
@@ -0,0 +1,737 @@
+//! would cause problems.
+//!
+//! Bandwidth scheduler handles this using a simple rule - if the last chunk was missing on some shard,
+//! don't send any receipts to this shard, it still hasn't processed the previous ones.
+//!
+//! TODO(bandwidth_scheduler) - think/prove/test if this is enough to handle missing chunks. We actually
+//! know if other chunks in the same block are missing, which is pretty powerful. I need to think more
+//! about it. Maybe there was no fatal flaw in the previous design after all? ;-;
+//!
+//! Future improvements: It's a bit wasteful to not send anything when a chunk is missing. We could track
+//! how much a shard already sent to the shard with missing chunks at previous heights, and allow to send
+//! receipts to it, but make sure that their total size doesn't exceed the grant that was given when the
+//! chunk wasn't missing.
+//! This would improve bandwidth utilization in scenarios with many missing chunks.
+
+use std::collections::{BTreeMap, VecDeque};
+use std::rc::Rc;
+
+use near_primitives::bandwidth_scheduler::{
+    Bandwidth, BandwidthRequest, BandwidthRequestValues, BandwidthRequests,
+    BandwidthSchedulerParams, BandwidthSchedulerState, BlockBandwidthRequests,
+};
+use near_primitives::types::ShardId;
+use rand::seq::SliceRandom;
+use rand::SeedableRng;
+use rand_chacha::ChaCha20Rng;
+
+use super::shard_mapping::{
+    SchedulerShardIndex, SchedulerShardIndexMap, SchedulerShardLinkMap, SchedulerShardMapping,
+    ShardLink,
+};
+
+/// How many bytes of outgoing receipts can be sent from one shard to another at the current height.
+/// Produced by the bandwidth scheduler.
+pub struct GrantedBandwidth {
+    granted: BTreeMap<(ShardId, ShardId), Bandwidth>,
+}
+
+impl GrantedBandwidth {
+    pub fn get_granted_bandwidth(&self, sender: ShardId, receiver: ShardId) -> Bandwidth {
+        self.granted.get(&(sender, receiver)).copied().unwrap_or(0)
+    }
+}
+
+pub struct BandwidthScheduler {
+    /// Mapping from shard ids to shard indexes.
+    mapping: SchedulerShardMapping,
+    /// List of shards that want to send receipts.
+    senders: Rc<[SchedulerShardIndex]>,
+    /// List of shards that the senders can send to.
+    receivers: Rc<[SchedulerShardIndex]>,
+    /// Configuration parameters for the algorithm.
+    params: BandwidthSchedulerParams,
+    /// ShardStatus for each shard.
+    shards_status: SchedulerShardIndexMap<ShardStatus>,
+    /// Each shard can send and receive at most `max_shard_bandwidth` bytes of receipts.
+    /// This is tracked in the `sender_budget` and `receiver_budget` fields, which keep
+    /// track of how much more a shard can send or receive before hitting the limit.
+    /// The sender and receiver budgets are decreased when bandwidth is granted on some link.
+    /// It's not possible to grant more bandwidth than the sender or receiver budget.
+    sender_budget: SchedulerShardIndexMap<Bandwidth>,
+    receiver_budget: SchedulerShardIndexMap<Bandwidth>,
```

Comment:
It is how much more bandwidth each shard can send at this height. It's the opposite of a grant, every time bandwidth is granted, the budget is decreased.
I could try to convert it to cumulative granted bandwidth, but I'll have to think about the implications. Originally there was an idea that if a shard shouldn't send/receive anything because of something, we could set its budget to zero and be sure that it won't be granted anything.
Another problem is that there could theoretically be bandwidth requests to a shard that isn't a receiver shard. With budgets this shard's `receiver_budget` will be zero. With cumulative grants we would have to check if the receiver shard is in the set of receivers.

Reply:
Eh I kinda don't want to rewrite everything with a different notation, I'll keep it as is unless someone is very against that.